pytorch_lightning/plugins/training_type/fully_sharded.py

```diff
@@ -0,0 +1,206 @@
+
+        Many of the helpful doc strings below came from the original FairScale documentation:
+        `https://fairscale.readthedocs.io/en/latest/api/nn/fsdp.html`
+
+        Arguments:
+            cpu_offload: Offload FP32 params to CPU. Only usable in precision=16 mode.
+                (Default: False).
+            move_grads_to_cpu: Moves gradient shards to CPU after reduction.
+                Only disable if using CPU based optimizers
+                (Default to ``cpu_offload``).
+            flatten_parameters: Flattens parameter into single contiguous tensor for speed efficiency
+                (Default: True).
+            reshard_after_forward: Reshard parameters after the forward pass, which saves memory but slows
+                down training. This is only relevant when resharding individual layers.
+                (Default: True).
+            fp32_reduce_scatter: Reduce-Scatter gradients in FP32. Only relevant in mixed precision
+                (Default: None).
+            compute_dtype: dtype for full parameters for computation. Default to torch.float32,
+                unless using mixed precision, in which case defaults to torch.float16.
+                (Default: None).
+            bucket_cap_mb: bucket parameters so that gradient reduction
+                can potentially overlap with backward computation.
+                bucket_cap_mb controls the bucket size in MegaBytes (MB).
+                Buckets are sub-divided based on world_size,
+                so the max shard size is roughly bucket_cap_mb / world_size.
+                Values <= 0 disable bucketing.
+                (Default: 25).
+            min_num_params: Number of parameters to wrap when using FairScale ``auto_wrap``.
+                (Default: 1e8)
+            state_dict_to_cpu: Whether to return parameters (returned by :func:`state_dict`) on CPU device.
+                If ``False``, this will default to ``compute_device``.
+                (Defautl: True).
+        """
+
+        super().__init__(
+            parallel_devices=parallel_devices,
+            cluster_environment=cluster_environment,
+        )
+        self.cpu_offload = cpu_offload
+        self.move_grads_to_cpu = move_grads_to_cpu
+        self.flatten_parameters = flatten_parameters
+        self.reshard_after_forward = reshard_after_forward
+        self.fp32_reduce_scatter = fp32_reduce_scatter
+        self.compute_dtype = compute_dtype
+        self.bucket_cap_mb = bucket_cap_mb
+        self.min_num_params = min_num_params
+        self.state_dict_device = torch.device("cpu") if state_dict_to_cpu else None
+        self._process_group = None
+
+    @property
+    def process_group(self):
+        if self._process_group is None:
+            self._process_group = torch.distributed.new_group()
+        return self._process_group
+
+    def setup_distributed(self):
+        super().setup_distributed()
+        if self.root_device.type == "cuda":
```

Comment:
does FSDP work for CPU only?

Reply:
it should not, cc @SeanNaren

Reply:
Haven't tested it, but iirc it relies on CUDA streams being available

Reply:
@shuyingsunshine21 in this case let's assert that since the DDP plugin this extends supports ddp_cpu. do you think that should happen here or in the accelerator connector? 

additionally, should the FSDP plugin  extend the parallel plugin instead of the DDP plugin? could we run into other issues if the ddp plugin is changed?

Reply:
> let's assert that since the DDP plugin this extends supports ddp_cpu. do you think that should happen here or in the accelerator connector

I think we could add assertion here. As we are thinking to flatten/merge accelerator/plugin, the AcceleratorStrategy should take care of this? wdyt?

Reply:
> should the FSDP plugin extend the parallel plugin instead of the DDP plugin? could we run into other issues if the ddp plugin is changed?

that is a good question, need to take a look